src/SparkTestBase.sol

```diff
@@ -40,23 +45,191 @@ interface IExecutable {
+    mapping(ChainId chainId => IExecutor executor) internal executors;
+    mapping(ChainId chainId => address payload)    internal payloads;
+    mapping(ChainId chainId => Domain domain)      internal domains;
+
+    Bridge internal baseBridge;
+
+    Domain[] internal allDomains;
+    string internal   id;
+
+    modifier onChain(ChainId chainId) {
+        ChainId currentChain = ChainIdUtils.fromUint(block.chainid);
+        domains[chainId].selectFork();
+        _;
+        domains[currentChain].selectFork();
+    }
+
+    /// @dev to be called in setUp
+    function setupDomains(uint256 mainnetForkBlock, uint256 baseForkBlock, uint256 gnosisForkBlock) internal {
+        domains[ChainIdUtils.Ethereum()] = getChain("mainnet").createFork(mainnetForkBlock);
+        domains[ChainIdUtils.Base()]     = getChain("base").createFork(baseForkBlock);
+        domains[ChainIdUtils.Gnosis()]   = getChain("gnosis_chain").createFork(gnosisForkBlock);
+
+        executors[ChainIdUtils.Ethereum()] = IExecutor(Ethereum.SPARK_PROXY);
+        executors[ChainIdUtils.Base()]     = IExecutor(Base.SPARK_EXECUTOR);
+        executors[ChainIdUtils.Gnosis()]   = IExecutor(Gnosis.AMB_EXECUTOR);
+
+        baseBridge = OptimismBridgeTesting.createNativeBridge(domains[ChainIdUtils.Ethereum()], domains[ChainIdUtils.Base()]);
+
+        allDomains.push(domains[ChainIdUtils.Ethereum()]);
+        allDomains.push(domains[ChainIdUtils.Base()]);
+        allDomains.push(domains[ChainIdUtils.Gnosis()]);
+    }
+
+    function spellIdentifier(ChainId chainId) private view returns(string memory){
+        string memory slug            = string(abi.encodePacked("Spark", chainId.toDomainString(), "_", id));
+        string memory spellIdentifier = string(abi.encodePacked(slug, ".sol:", slug));
+        return spellIdentifier;
+    }
+
+    function deployPayload(ChainId chainId) internal onChain(chainId) returns(address) {
+        return deployCode(spellIdentifier(chainId));
+    }
+
+    function deployPayloads() internal {
+        for (uint256 i = 0; i < allDomains.length; i++) {
+            ChainId id = ChainIdUtils.fromDomain(allDomains[i]);
+            string memory spellIdentifier = spellIdentifier(id);
+            try vm.getCode(spellIdentifier) {
+                payloads[id] = deployPayload(id);
+            } catch {
+                console.log("skipping spell deployment for network: ", id.toDomainString());
+            }
+        }
+    }
+
+    /// @dev takes care to revert the selected fork to what was chosen before
+    function executeAllPayloadsAndBridges() internal {
+        // only execute mainnet payload
+        executePayload(ChainIdUtils.Ethereum());
+        // then use bridges to execute other chains' payloads
+        relayMessageOverBridges();
+    }
+
+    /// @dev bridge contracts themselves are stored on mainnet
+    function relayMessageOverBridges() private onChain(ChainIdUtils.Ethereum()) {
```

Comment:
I would make this generic. Have a flag in the chain struct which says what type of domain it is: `optimism`, `arbitrum`, `gnosis`, etc and then use the corresponding Bridge Testing library to relay the messages.

Reply:
implemented in a91a65b